compiler/rustc_builtin_macros/src/assert.rs

```diff
@@ -23,16 +23,34 @@ pub fn expand_assert<'cx>(
         }
     };
 
+    let is_2021 = sp.rust_2021();
```

Comment:
Edition checks are generally expensive, so it's better to not perform them unless the result is actually used.
So it's better to move this to inside the `custom_message` condition (note that edition checking should use the `sp` before `with_call_site_ctxt`).